dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/TimeoutFilter.java

```diff
@@ -36,21 +36,34 @@
 
     private static final Logger logger = LoggerFactory.getLogger(TimeoutFilter.class);
 
+    private static final String TIMEOUT_FILTER_START_TIME = "timeout_filter_start_time";
+
     @Override
     public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcException {
-        long start = System.currentTimeMillis();
-        Result result = invoker.invoke(invocation);
-        long elapsed = System.currentTimeMillis() - start;
-        if (invoker.getUrl() != null
-                && elapsed > invoker.getUrl().getMethodParameter(invocation.getMethodName(),
-                "timeout", Integer.MAX_VALUE)) {
-            if (logger.isWarnEnabled()) {
-                logger.warn("invoke time out. method: " + invocation.getMethodName()
-                        + " arguments: " + Arrays.toString(invocation.getArguments()) + " , url is "
-                        + invoker.getUrl() + ", invoke elapsed " + elapsed + " ms.");
+        if (invocation.getAttachments() != null) {
+            long start = System.currentTimeMillis();
+            invocation.getAttachments().put(TIMEOUT_FILTER_START_TIME, String.valueOf(start));
+        }
+        return invoker.invoke(invocation);
+    }
+
+    @Override
+    public Result onResponse(Result result, Invoker<?> invoker, Invocation invocation) {
+        String startAttach = invocation.getAttachment(TIMEOUT_FILTER_START_TIME);
+        if (startAttach != null) {
+            long elapsed = System.currentTimeMillis() - Long.valueOf(startAttach);
+            if (invoker.getUrl() != null
+                    && elapsed > invoker.getUrl().getMethodParameter(invocation.getMethodName(),
+                    "timeout", Integer.MAX_VALUE)) {
+                if (logger.isWarnEnabled()) {
+                    logger.warn("invoke time out. method: " + invocation.getMethodName()
+                            + " arguments: " + Arrays.toString(invocation.getArguments()) + " , url is "
+                            + invoker.getUrl() + ", invoke elapsed " + elapsed + " ms.");
+                }
             }
+            // we should not change the attachments of the invocation.
+            invocation.getAttachments().remove(TIMEOUT_FILTER_START_TIME);

```

Comment:
I am not sure should we care about the scenario here, where in case of exception while executing _invoker.invoke(invocation)_ , should we remove TIMEOUT_FILTER_START_TIME from attacchment ? I think as this piece of code should take of removing TIMEOUT_FILTER_START_TIME even in case of exception. What do you say?

Reply:
Agree.
I will fix it.

Reply:
Hi, @khanimteyaz 
After check the code, seems like it would throw any exception..
So I keep this code.

Reply:
> Hi, @khanimteyaz
> After check the code, seems like it would throw any exception..
> So I keep this code.

In this case would we be missing the chance to remove TIMEOUT_FILTER_START_TIME  from attachment?

Reply:
Won't miss remove TIMEOUT_FILTER_START_TIME unless there is a bug.
But in bug case, we need fix our code but not add a try catch..I think.
What do u  think about it?
:)

Reply:
If there are bug we any how need to fix it :). 

I think it is fine, as I was worring about whether we are missing to remove it or not, but on our known path we are removing.

I have created a issue to write a UT for specific this scenario and will take it 
https://github.com/apache/incubator-dubbo/issues/3234